slsDetectorServers/slsDetectorServer/src/sharedMemory.c

```diff
@@ -0,0 +1,200 @@
+#include "sls_detector_defs.h"
+
+#include <errno.h> // errno
+#include <fcntl.h> // O_CREAT, O_TRUNC..
+#include <string.h>
+#include <sys/mman.h> // shared memory
+#include <sys/stat.h> // fstat
+#include <unistd.h>
+
+#include <sys/ipc.h>
+#include <sys/shm.h>
+#include <sys/types.h>
+
+#define SHM_NAME    "sls_server_shared_memory"
+#define SHM_VERSION 0x200625
+#define SHM_KEY     5678
+
+typedef struct Memory {
+    int version;
+    sem_t sem;
+    int scanStatus;
+    int scanStop;
+#ifdef VIRTUAL
+    int status;
+    int stop;
+#endif
+} sharedMem;
+
+sharedMem *shm = NULL;
+char shmMess[MAX_STR_LENGTH];
+int shmFd = -1;
+
+extern int isControlServer;
+
+char *sharedMemory_getError() { return shmMess; }
+
+void sharedMemory_print() {
+    LOG(logINFO, ("%s Shared Memory:\n", isControlServer ? "c" : "s"));
+    LOG(logINFO,
+        ("%s version:0x%x\n", isControlServer ? "c" : "s", shm->version));
+#ifdef VIRTUAL
+    LOG(logINFO, ("%s status: %d\n", isControlServer ? "c" : "s", shm->status));
+    LOG(logINFO, ("%s stop: %d\n", isControlServer ? "c" : "s", shm->stop));
+#endif
+}
+
+int sharedMemory_create(int port) {
+    memset(shmMess, 0, MAX_STR_LENGTH);
+
+    // if sham existed, delete old shm and create again
+    shmFd =
+        shmget(SHM_KEY + port, sizeof(sharedMem), IPC_CREAT | IPC_EXCL | 0666);
+    if (shmFd == -1 && errno == EEXIST) {
+        LOG(logWARNING, ("Removing old shared memory\n"));
+        // cuz of unknown previous shm size, it has to  be deleted like this
+        system(
+            "for i in seq `ipcs -m | cut -d ' ' -f1`; do ipcrm -M $i; done;");
+        shmFd = shmget(SHM_KEY + port, sizeof(sharedMem),
+                       IPC_CREAT | IPC_EXCL | 0666);
+    }
+    if (shmFd == -1) {
+        sprintf(shmMess, "Create shared memory failed: %s\n", strerror(errno));
+        LOG(logERROR, (shmMess));
+        return 0;
+    }
+    LOG(logINFO, ("Shared memory created\n"));
+    if (!sharedMemory_attach()) {
+        return 0;
+    }
+    sharedMemory_initialize();
+    return 1;
+}
+
+void sharedMemory_initialize() {
+    shm->version = SHM_VERSION;
+    sem_init(&(shm->sem), 1, 1);
+#ifdef VIRTUAL
+    shm->status = 0;
+    shm->stop = 0;
+#endif
+    LOG(logINFO, ("Shared memory initialized\n"))
+}
+
+int sharedMemory_open(int port) {
+    memset(shmMess, 0, MAX_STR_LENGTH);
+    shmFd = shmget(SHM_KEY + port, sizeof(sharedMem), 0666);
+    if (shmFd == -1) {
+        sprintf(shmMess, "Open shared memory failed: %s\n", strerror(errno));
+        LOG(logERROR, (shmMess));
+        return 0;
+    }
+    if (!sharedMemory_attach()) {
+        return 0;
+    }
+    if (shm->version != SHM_VERSION) {
+        sprintf(shmMess,
+                "Shared memory version 0x%x does not match! (expected: 0x%x)\n",
+                shm->version, SHM_VERSION);
+        LOG(logERROR, (shmMess));
+    }
+    LOG(logINFO, ("Shared memory opened\n"));
+    return 1;
+}
+
+int sharedMemory_attach() {
```

Comment:
Its not difficult. I put this in because everywhere for Ians part of Eiger, its done other way around.. so I started using that whereever we dont have OK or FAIL. Ill switch to OK and FAIL since we already include sls_detector_defs